internal/installercache/installercache.go

```diff
@@ -1,61 +1,160 @@
 package installercache
 
 import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
 	"sync"
+	"time"
 
 	"github.com/openshift/assisted-service/internal/oc"
 	"github.com/openshift/assisted-service/models"
-	"github.com/openshift/assisted-service/pkg/executer"
-	"github.com/openshift/assisted-service/pkg/mirrorregistries"
+	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
-type installers struct {
+var GRACE_PERIOD time.Duration = -5 * time.Minute
+
+type Installers struct {
 	sync.Mutex
-	releases map[string]*release
+	log      logrus.FieldLogger
+	capacity int
+	cacheDir string
 }
 
-type release struct {
-	sync.Mutex
+type fileInfo struct {
 	path string
+	info os.FileInfo
 }
 
-var cache installers = installers{
-	releases: make(map[string]*release),
+func (fi *fileInfo) Compare(other *fileInfo) bool {
+	//oldest file will be first in queue
+	return fi.info.ModTime().Unix() < other.info.ModTime().Unix()
 }
 
-// Get returns a release resource for the given release ID
-func (i *installers) Get(releaseID string) *release {
-	i.Lock()
-	defer i.Unlock()
-
-	r, present := i.releases[releaseID]
-	if !present {
-		r = &release{}
-		i.releases[releaseID] = r
+func New(cacheDir string, capacity int, log logrus.FieldLogger) *Installers {
+	return &Installers{
+		log:      log,
+		capacity: capacity,
+		cacheDir: cacheDir,
 	}
-	return r
 }
 
 // Get returns the path to an openshift-baremetal-install binary extracted from
 // the referenced release image. Tries the mirror release image first if it's set. It is safe for concurrent use. A cache of
 // binaries is maintained to reduce re-downloading of the same release.
-func Get(releaseID, releaseIDMirror, cacheDir, pullSecret string, platformType models.PlatformType, icspFile string, log logrus.FieldLogger) (string, error) {
-	r := cache.Get(releaseID)
-	r.Lock()
-	defer r.Unlock()
+func (i *Installers) Get(releaseID, releaseIDMirror, pullSecret string, ocRelease oc.Release, platformType models.PlatformType, icspFile string) (string, error) {
+	i.Lock()
+	defer i.Unlock()
 
-	var path string
+	var workdir, binary, path string
 	var err error
-	//cache miss
-	if r.path == "" {
-		mirrorRegistriesBuilder := mirrorregistries.New()
-		path, err = oc.NewRelease(&executer.CommonExecuter{}, oc.Config{
-			MaxTries: oc.DefaultTries, RetryDelay: oc.DefaltRetryDelay}, mirrorRegistriesBuilder).Extract(log, releaseID, releaseIDMirror, cacheDir, pullSecret, platformType, icspFile)
+
+	workdir, binary, path = ocRelease.GetReleaseBinaryPath(releaseID, i.cacheDir, platformType)
+	if _, err = os.Stat(path); os.IsNotExist(err) {
+		//evict older files if necessary
+		i.evict()
+
+		//extract the binary
+		_, err = ocRelease.Extract(i.log, releaseID, releaseIDMirror, i.cacheDir, pullSecret, platformType, icspFile)
 		if err != nil {
 			return "", err
 		}
-		r.path = path
+	} else {
+		//update the file mtime to signal it was recently used
```

Comment:
nit: shouldn't we only change ``atime``? because if we only accessed the file (and not modified it) the ``mtime`` isn't expected to change in linux

Reply:
`atime` is not updated for every access on most linux systems, so it's not a reliable or accurate depiction of when a file was used. You can imagine the performance impact if every file access included a write to disk, however small. See the `relatime` mount option for details.

Reply:
yes, I guessed atime is not updated on any change in linux
but my point is about mtime not being updated for any read action

Reply:
any way it's not critical. but if a means "accessed" and m means "modified" it seems more reasonable to only update ``atime`` and leave ``mtime`` updated each time the file actually changes